extension/src/content_scripts/google_classroom/FacialDetection.js

```diff
@@ -27,32 +27,64 @@ async function renderPrediction() {
                     predictions[i].landmarks = predictions[i].landmarks.arraySync();
                 }
             }
-  
-            const start = predictions[i].topLeft;
-            const end = predictions[i].bottomRight;
-            const size = [end[0] - start[0], end[1] - start[1]];
-            GVars.ctx.strokeStyle = "white";
-            GVars.ctx.strokeRect(start[0], start[1], size[0], size[1]);
-  
-            if (annotateBoxes) {
-                const landmarks = predictions[i].landmarks;
-                GVars.ctx.fillStyle = "white";
-                for (let j = 0; j < landmarks.length; j++) {
-                    const x = landmarks[j][0];
-                    const y = landmarks[j][1];
-                    GVars.ctx.fillRect(x, y, 5, 5);
-                }
+
+            // Convert Image to tensor
+            const faceInput = tf.browser.fromPixels(GVars.video, 3)
+                .expandDims(0);   
+            const input = tf.tensor4d(Array.from(faceInput.dataSync()), [1, faceInput.shape[1], faceInput.shape[2], 3]);
+
+            // crop face based on facial detection results
+            const normalizeStart = tf.div(predictions[i].topLeft, input.shape.slice([1], [2]));
+            const normalizeEnd = tf.div(predictions[i].bottomRight, input.shape.slice([1], [2]));
+            const boxes = tf.concat([normalizeStart, normalizeEnd]).reshape([-1, 4]);
+            const cropFace = tf.image.cropAndResize(faceInput, boxes, [0], [150, 150]);
+
+            // preprocessing grayscale
+            const r = cropFace.slice([0, 0, 0, 0], [1, cropFace.shape[1], cropFace.shape[2], 1]);
+            const b = cropFace.slice([0, 0, 0, 1], [1, cropFace.shape[1], cropFace.shape[2], 1]);
+            const g = cropFace.slice([0, 0, 0, 2], [1, cropFace.shape[1], cropFace.shape[2], 1]);
+            const avg = tf.div(tf.addN([r, b, g]), tf.tensor([3]));
+            const grayInput = tf.concat([avg, avg, avg], 3);
+            
+            // expression prediction
+            let prediction = await GVars.facialExpressionModel.predict(grayInput);
+            console.log(prediction.dataSync()[0]);
+            let confidence = Math.abs(prediction.dataSync()[0] - .5) / .5;
+            let confidenceLabel = confidence < .5 ? true : false;
+            let label = prediction.dataSync()[0] < .5 ? true : false;
+            console.log(label);
+            console.log(confidenceLabel);
+
+            // TODO: send passive signal to backend
+            // remember to add time check so it doesnt overflow the backend
+
+            if (label) {
```

Comment:
this is rather entertaining to look at 2AM. I hope you implemented it this way because something went wrong in the issue transition process and some context got lost.

The issue with this implementation is that, if you make a distressed face for a few seconds. Hundreds of requests will hit the EC2 instance. I am not looking to upgrade the instance due to budget constraint.

If you believe the model is very slow, please record and capture a trace of the network pattern in the chrome devtool. I would like avoid a AWS bill at all cost 😶. 

I would prefer to see a better rate limiting implementation than a wide open gate like this. A token bucket / leaky bucket would be great here. Good news, it is available [here](https://www.npmjs.com/package/leaky-bucket). If you want to use it, let me know and I will install it for you since your `npm` is producing a different version of `package-lock`.

Also, please address my question in #82. What does a unit (a counter in the Y value) in the passive graph represent with this implementation? A count represent a student is confused at time `x`? Since you are the E2E feature owner now? How do you envision the graph feature (values and what they represent)?

if we manually shape the traffic via something like leaky bucket above, you want to adjust how you treat the timestamp (tho not 100% necessary). For example, if we allow one passive signal per second, then you want to cast the return value of `GVars.player.getCurrentTime();` into an integer.

I don't know if this is related, but just FYI. Time returned from the player can be highly granular (11.213213). However, only the `integer` part matter for keyword extraction. I don't know if a float is accepted by the python backend.

Reply:
The renderPrediction function is not called per frame.
The model is very slow and is performed synchronously to the next possible call.

What does means is if renderPrediction is called on timeframe 1 and takes 12 seconds to complete, the next call to renderPrediction will be after 12 seconds. Even then, there won't be a post to the backend if no emotion is detected.

I'm fine with implementing a leaky-bucket just know that it won't have an effective difference on the number of post requests. It'll only matter if we somehow get a machine with a massive GPU suddently. It'll just be for peace of mind.

Reply:
I am aware that renderPrediction is not called per frame since the v8 runtime only do things in the best-effort way. However, I am not convinced by the claim that the model is very slow without any numbers. One reason is that I deprecated CPU and used WebAssembly for model execution.

 Leaky bucket has been installed. Please use it to throttle calls within `sendSignal`. 


Maybe I missed a message from you? I am assuming that you are sticking with floating point timestamps for passive signals?

Does the dashboard render the passive signal after you? A screenshot of UI illustrating the E2E flow would be great

Reply:
I forgot to address the floating point. I can truncate timestamps if needed.

Reply:
A screenshot wouldn't really do much for E2E so I want to do a short video.